### PR TITLE
feat(slice-4b): streaming llm adapter + dec-073 stream-error translation (pr1)

### PR DESCRIPTION
**Pull request by leehopper**
## Slice 4B — PR1 (Unit 1): Streaming LLM adapter + DEC-073 stream-error translation

First piece of the streaming conversation core (DEC-085). Adds `ICoachingLlm.StreamAsync` — a net-new streaming method over the Anthropic 12.29.1 + Microsoft.Extensions.AI streaming bridge — and translates the streaming failure surface (R-084) into the DEC-073 totality contract. No orchestration consumes it yet (PR4 wires the SSE endpoint); this is the foundational adapter and lands behind an unreferenced method.

### What's in it
- **`ICoachingLlm.StreamAsync`** (`IAsyncEnumerable<string>` text deltas) + the net-new **`IncompleteCoachingLlmException`** / `IncompleteReason` errored-turn contract.
- **`ClaudeCoachingLlm.StreamAsync`** drives the bridge via a manual enumerator (`MoveNextAsync` in try/catch, `yield` outside), routed through `SanitizationAuditChatClient` for the GUARDRAIL span; argument validation is eager (split into a private `StreamCore` iterator, sonar S4456).
- **Totality contract** — only `Transient` / `Permanent` / `Incomplete` CoachingLlmException, plus an unwrapped `OperationCanceledException` on a genuine client abort, can escape:
  - mid-stream `AnthropicSseException` → classified on the SDK's **typed `ErrorType` enum** (rate-limit / overloaded / api / timeout / unknown → Transient; invalid-request / auth / permission / not-found / billing → Permanent);
  - pre-first-byte `AnthropicApiException` → the shared request/response status classifier (408/409/429/5xx → Transient, else Permanent), with the pre-stream 429 `Retry-After` captured via the owned pipeline;
  - `max_tokens` / `model_context_window_exceeded` / `refusal` end the stream **cleanly** → `IncompleteCoachingLlmException` with the right `Reason` + `Retryable`, detected from the raw `stop_reason` (the bridge maps context-window to `FinishReason.Stop`);
  - client abort propagates unwrapped (not-an-error); the SDK per-attempt timeout → Transient.
- **DRY:** `CreateMessageAsync` and `StreamAsync` now share one `TranslateAnthropicFailure` classifier.

### Verification
- R-084's open caveat resolved **empirically** against the real SDK (the inherited `AnthropicServiceException.ErrorType` accessor + the context-window→`Stop` mapping), not from docs.
- **TDD throughout:** stub-`IChatClient` translator unit tests + real-SDK-over-SSE-stub transport tests (mid-stream classification incl. message-text independence, incomplete finishes, pre-first-byte status + Retry-After, abort, eager arg validation, GUARDRAIL audit-span wiring).
- An adversarial multi-lens self-review (find → blind-challenge) ran on the diff; its 6 confirmed findings are all addressed in `342205f` (typed `ErrorType` classification, eager-validation split, `ValueKind` guard, partial-deltas-before-throw tests, AsyncLocal-scope clarification).
- **Full backend suite 1835/1835 green in Replay** (integration via Testcontainers).

Per the stacked-train plan this is a root PR (PR2 is the parallel root; PR4 is the integration gate that first wires `StreamAsync`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added `StreamAsync` for coaching to stream text deltas as they arrive.
  * Introduced `IncompleteCoachingLlmException` and `IncompleteReason` to report when a stream ends without a usable complete reply.
* **Bug Fixes**
  * Improved streaming error handling with correct transient vs permanent classification, proper propagation of cancellation, and preservation of already-received deltas.
  * Enhanced trademark scrubbing to correctly redact terms even when split across streamed chunks.
* **Tests**
  * Added coverage for streaming yields, eager validation, cancellation behavior, transport/SSE failure classification, incomplete-stop handling, and scrub boundary edge cases.

### CODE REVIEW

File: backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs [hunk 1]
```diff
@@ -455,6 +490,113 @@ private static string ScrubTrademarkedProse(ILogger logger, string text, string
...
+
+        if (reason is not { } incompleteReason)
+        {
+            return;
+        }
+
+        throw new IncompleteCoachingLlmException(
+            IncompleteMessage,
+            incompleteReason,
+            retryable: incompleteReason == IncompleteReason.MaxTokens);
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs]: **💡 [LOW] Feature file contradicts throw-based errored-turn signal**

The `streaming-adapter.feature` scenario "A max_tokens finish surfaces an errored-turn signal, not an exception" has **"Then no exception is thrown"** (twice), but this throws `IncompleteCoachingLlmException`. The PR updated the `ICoachingLlm.StreamAsync` doc to make `IncompleteCoachingLlmException` part of the totality contract (a reasonable design), but left the feature file contradicting it — the sibling scenario lists only Transient/Permanent as escaping. **Fix:** update the feature scenario to assert `IncompleteCoachingLlmException` is thrown carrying reason + retryability.

File: backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs [hunk 2]
```diff
@@ -455,6 +490,113 @@ private static string ScrubTrademarkedProse(ILogger logger, string text, string
...
+    /// <c>Retry-After</c> header once SSE headers are flushed, so a transient mid-stream error
+    /// carries no delay hint (the SSE endpoint applies a configured default).
+    /// </summary>
+    private static CoachingLlmException ClassifySseError(AnthropicSseException sse) => sse.ErrorType switch
+    {
+        ErrorType.InvalidRequestError or ErrorType.AuthenticationError or ErrorType.PermissionError
+            or ErrorType.NotFoundError or ErrorType.BillingError
+            => new PermanentCoachingLlmException(RejectedMessage, sse),
+        _ => new TransientCoachingLlmException(UnavailableMessage, retryAfterSeconds: null, sse),
+    };
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs:632]: **💡 [LOW] Mid-stream Transient RetryAfterSeconds spec divergence**

The feature file + spec doc line 51 say a mid-stream Transient "carries a configured default RetryAfterSeconds", but `ClassifySseError` passes `retryAfterSeconds: null` and defers the default to "the SSE endpoint" — which does not exist yet, and no configured-default setting exists. The clause is also untested. **Fix:** either update the spec to say the exception carries `null` and PR4 (the SSE endpoint) owns the default, or wire a `CoachingLlmSettings` default now. Updating the spec is likely cleaner since PR4 is the integration gate.

File: backend/tests/RunCoach.Api.Tests/Modules/Coaching/ClaudeCoachingLlmStreamingTransportTests.cs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        var act = () => DrainIntoAsync(llm, deltas);
+
+        // Assert — the partial reached the consumer (so the caller can discard it per the contract),
+        // then an IncompleteCoachingLlmException carrying the precise reason and retryability, never
+        // a clean completion.
+        var thrown = await act.Should().ThrowAsync<IncompleteCoachingLlmException>();
+        thrown.Which.Reason.Should().Be(expectedReason);
+        thrown.Which.Retryable.Should().Be(expectedRetryable);
+        deltas.Should().Equal("Easy ", "does it.");
+    }
```
Comment (leehopper) [inline backend/tests/RunCoach.Api.Tests/Modules/Coaching/ClaudeCoachingLlmStreamingTransportTests.cs:132]: **💡 [LOW] No test covers incomplete finish with zero text deltas**

`ThrowIfIncompleteFinish` runs unconditionally, but every incomplete-finish test prepends two deltas via `SseBuilder.AppendPreamble`. A realistic Anthropic case (`max_tokens`/`refusal` with no emitted text) is untested — a refactor gating the throw on "a delta was yielded" would pass all existing tests while silently dropping the incomplete signal for empty truncations. **Fix:** add a transport test whose SSE body emits a `max_tokens` stop_reason with zero `content_block_delta` events; assert `IncompleteCoachingLlmException` + empty deltas.

File: backend/src/RunCoach.Api/Modules/Coaching/IncompleteCoachingLlmException.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        Reason = reason;
+        Retryable = retryable;
+    }
+
+    /// <summary>Gets the reason the reply did not complete normally.</summary>
+    public IncompleteReason Reason { get; }
+
+    /// <summary>Gets a value indicating whether re-sending the turn could plausibly succeed.</summary>
+    public bool Retryable { get; }
+}
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/IncompleteCoachingLlmException.cs:47]: **💡 [LOW] Constructor allows Reason/Retryable mismatch (unenforced invariant)**

The ctor takes `reason` and `retryable` independently, so `new IncompleteCoachingLlmException(msg, IncompleteReason.Refusal, retryable: true)` compiles despite the documented invariant (retryable only for `MaxTokens`). The sole construction site is correct, so no live bug — a latent API-design weakness. **Fix:** drop the `retryable` parameter and derive it in the ctor: `Retryable = reason == IncompleteReason.MaxTokens;`.

```suggestion
public IncompleteCoachingLlmException(
    string message,
    IncompleteReason reason,
    Exception? innerException = null)
    : base(message, innerException)
{
    Reason = reason;
    Retryable = reason == IncompleteReason.MaxTokens;
}
```

**Review by leehopper — COMMENTED** [review]
## Deep Review — PR #219 (streaming LLM adapter + DEC-073 stream-error translation)

Strong, TDD-driven PR. Bug & security review found nothing; the `TranslateAnthropicFailure` DRY refactor was verified behaviorally equivalent. **6 findings, all low severity**, all confirmed by blind challenge.

Notably, the two spec-divergence findings (F-01, F-02) were initially dismissed by validators but **rescued by the blind challengers** who read the feature file — they are the most actionable. None are merge-blockers.

*Generated by deep-review · Frontier mode · Reviewed up to: 342205fe*

File: backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs [hunk 5]
```diff
@@ -486,20 +628,102 @@ private static void ValidateSettings(CoachingLlmSettings settings)
...
+                }
+
+                // The terminal stop reason rides the message_delta update (and may repeat on
+                // trailing updates — last-write-wins keeps that idempotent). Captured here, acted
+                // on after the stream ends cleanly.
+                var stopReason = TryGetTerminalStopReason(update);
+                if (stopReason is not null)
+                {
+                    terminalStopReason = stopReason;
+                }
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs]: **💡 [LOW] Simplify: if-block to null-coalescing assignment**

Per backend CLAUDE.md (compact conditional assignment), the `stopReason` if-block can collapse to a single line. Behaviorally identical (single evaluation, last-write-wins). Keep the existing comment.

```suggestion
            terminalStopReason = TryGetTerminalStopReason(update) ?? terminalStopReason;
```

**Comment by leehopper** [comment]
**Deep Review — F-03 (low): stale base-class doc** (couldn't attach inline — `CoachingLlmException.cs` isn't in this PR's diff)

`backend/src/RunCoach.Api/Modules/Coaching/CoachingLlmException.cs` line 7 still says the adapter translates failures "into one of the **two** concrete subtypes," but this PR adds a third (`IncompleteCoachingLlmException`) and updated the `ICoachingLlm.StreamAsync` doc to list all three — leaving the base-class doc internally inconsistent within the same changeset. A future reader may write a two-arm catch and miss the third type.

**Fix:** update the summary to enumerate all three subtypes (Transient / Permanent / Incomplete).

*Generated by deep-review*

File: backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs [hunk 6]
```diff
@@ -486,20 +628,102 @@ private static void ValidateSettings(CoachingLlmSettings settings)
...
+                if (!string.IsNullOrEmpty(delta))
+                {
+                    yield return delta;
+                }
+            }
+
+            // A clean enumeration end on an incomplete stop reason (truncation, context overflow,
+            // or a refusal) is the errored-turn signal: the partial just yielded is unusable as a
+            // complete turn. R-084 — these end the stream cleanly, never as an exception.
+            ThrowIfIncompleteFinish(terminalStopReason);
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs:773]: Addressed in ea90fcd. `StreamCore` now harvests `input_tokens` (+ model id) from the `message_start` event and the cumulative `output_tokens` from `message_delta` (new `HarvestStreamUsage` helper), then calls the existing `LogReceivedResponse` once the stream ends — before the errored-turn check so telemetry is captured even on an incomplete finish. Cost-tracking parity with the non-streaming path restored.

File: backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs [hunk 7]
```diff
@@ -486,20 +628,102 @@ private static void ValidateSettings(CoachingLlmSettings settings)
...
+                if (stopReason is not null)
+                {
+                    terminalStopReason = stopReason;
+                }
+
+                var delta = update.Text;
+                if (!string.IsNullOrEmpty(delta))
+                {
+                    yield return delta;
+                }
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs]: Addressed in ea90fcd. Added `StreamingTrademarkScrubber` and wired it into `StreamCore`, so each delta is scrubbed before it is yielded — the same boundary `GenerateAsync` enforces on complete responses. It handles the split-across-deltas case by holding back only the trailing run that could still complete into the term (`V[.-]?DOTs?`), so clean streams keep their natural chunking and incur no extra latency. Covered by `StreamingTrademarkScrubberTests` (split-term cases) plus a stream-level `StreamAsync_ScrubsTrademarkedTerm_EvenWhenSplitAcrossDeltas` transport test.

File: backend/src/RunCoach.Api/Modules/Coaching/ICoachingLlm.cs [hunk 8]
```diff
@@ -77,4 +77,30 @@ public interface ICoachingLlm
         IReadOnlyDictionary<string, JsonElement>? schema,
         CacheControl? cacheControl,
         CancellationToken ct);
+
+    /// <summary>
+    /// Streams the LLM's free-text reply token-by-token for the interactive
+    /// coaching conversation (Slice 4B). Yields each text delta as it arrives.
+    /// </summary>
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/ICoachingLlm.cs:84]: Done in ea90fcd — dropped the `(Slice 4B)` reference and replaced `token-by-token` with `as text deltas` to match the provider-level behaviour.

File: backend/src/RunCoach.Api/Modules/Coaching/IncompleteReason.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+namespace RunCoach.Api.Modules.Coaching;
+
+/// <summary>
+/// Why a streamed coaching reply ended without a usable, complete turn
+/// (Slice 4B / R-084). The SDK reports each of these as a clean enumeration end
+/// rather than an exception, so <see cref="ICoachingLlm.StreamAsync"/> detects the
+/// terminal finish reason and raises <see cref="IncompleteCoachingLlmException"/>.
+/// Explicitly numbered for stable wire/telemetry encoding.
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/IncompleteReason.cs:8]: Done in ea90fcd — removed `Slice 4B /`, kept the durable `R-084` marker.

File: backend/tests/RunCoach.Api.Tests/Infrastructure/StubCoachingLlm.cs [hunk 10]
```diff
@@ -80,6 +80,13 @@ public Task<string> GenerateAsync(string systemPrompt, string userMessage, Cance
             "StubCoachingLlm.GenerateAsync was called: no integration-tier flow scripts free-text "
             + "generation. Script the structured surface instead, or move the coverage to the eval tier.");
 
+    /// <inheritdoc />
+    public IAsyncEnumerable<string> StreamAsync(string systemPrompt, string userMessage, CancellationToken ct) =>
+        throw new InvalidOperationException(
+            "StubCoachingLlm.StreamAsync was called: no integration-tier flow scripts streaming "
+            + "generation yet. Streaming unit coverage lives in ClaudeCoachingLlmStreamingTests; "
+            + "end-to-end conversation coverage lands with the Slice 4B SSE endpoint.");
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/tests/RunCoach.Api.Tests/Infrastructure/StubCoachingLlm.cs]: Done in ea90fcd — removed the future-SSE-endpoint wording; the diagnostic still points at the `ClaudeCoachingLlm` streaming tests and suggests a dedicated streaming test seam.

File: backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs (hunk 1 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs]: Addressed in the working-tree feature file (ea90fcd touches source only — `docs/specs/` is gitignored, so it is not part of the PR diff). The `max_tokens`/context-window/refusal scenarios now assert `IncompleteCoachingLlmException` is thrown carrying the matching `Reason` + `Retryable`, consistent with the totality contract on `ICoachingLlm.StreamAsync`.

File: backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs (hunk 2 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs:632]: Updated the spec doc (line 51) and feature file in the working tree (gitignored, not in the PR diff): a mid-stream Transient now documents `RetryAfterSeconds = null` from the adapter, with PR4's SSE endpoint owning the configured default. Matches `ClassifySseError` passing `retryAfterSeconds: null`.

File: backend/tests/RunCoach.Api.Tests/Modules/Coaching/ClaudeCoachingLlmStreamingTransportTests.cs (hunk 3 above)
Comment (leehopper) [reply backend/tests/RunCoach.Api.Tests/Modules/Coaching/ClaudeCoachingLlmStreamingTransportTests.cs:132]: Done in ea90fcd. Added `StreamAsync_IncompleteFinishWithNoTextDeltas_StillRaisesIncomplete` (transport tier): an SSE body emitting a `max_tokens`/`refusal` stop_reason with zero `content_block_delta` events asserts `IncompleteCoachingLlmException` + empty deltas — pins that the errored-turn signal is delta-independent.

File: backend/src/RunCoach.Api/Modules/Coaching/IncompleteCoachingLlmException.cs (hunk 4 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/IncompleteCoachingLlmException.cs:47]: Done in ea90fcd. Dropped the `retryable` parameter; the ctor now derives `Retryable = reason == IncompleteReason.MaxTokens`, making the invariant unbreakable at construction. Updated the sole call site.

File: backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs (hunk 5 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/ClaudeCoachingLlm.cs]: Done in ea90fcd — collapsed to `terminalStopReason = TryGetTerminalStopReason(update) ?? terminalStopReason;`, comment retained.

**Comment by leehopper** [comment]
Addressed in d23c371. `CoachingLlmException`'s summary now enumerates all three subtypes — `TransientCoachingLlmException` / `PermanentCoachingLlmException` / `IncompleteCoachingLlmException` — so the base-class doc no longer says "two" and is consistent with the updated `ICoachingLlm.StreamAsync` contract.
